modulemd/modulemd-eol.c

```diff
@@ -0,0 +1,1207 @@
+ *
+ * This program is free software.
+ * For more information on the license, see COPYING.
+ * For more information on free software, see <https://www.gnu.org/philosophy/free-sw.en.html>.
+ */
+
+#include <glib.h>
+#include <inttypes.h>
+#include <yaml.h>
+
+#include "modulemd-errors.h"
+#include "modulemd-eol.h"
+#include "private/modulemd-util.h"
+#include "private/modulemd-eol-private.h"
+
+#include "private/modulemd-subdocument-info-private.h"
+#include "private/modulemd-util.h"
+#include "private/modulemd-yaml.h"
+
+#define O_DEFAULT_STRING "__eol_VALUE_UNSET__"
+#define O_PLACEHOLDER_STRING "__eol_VALUE_NOT_YET_SET__"
+
+struct _ModulemdEol
+{
+  GObject parent_instance;
+
+  guint64 mdversion;
+  guint64 modified;
+  gboolean reset;
+  gchar *module_name;
+  gchar *module_stream;
+  gchar *module_context;
+  guint64 eol_date;
+  gchar *message;
+
+  /* Stream is obsoleted by exactly one other stream */
+  gchar *obsoleted_by_module_name;
+  gchar *obsoleted_by_module_stream;
+};
+
+G_DEFINE_TYPE (ModulemdEol, modulemd_eol, G_TYPE_OBJECT)
+
+enum
+{
+  PROP_0,
+
+  PROP_MDVERSION,
+  PROP_MODIFIED,
+  PROP_RESET,
+  PROP_MODULE_NAME,
+  PROP_MODULE_STREAM,
+  PROP_MODULE_CONTEXT,
+  PROP_EOL_DATE,
+  PROP_MESSAGE,
+  PROP_OBSOLETED_BY_MODULE_NAME,
+  PROP_OBSOLETED_BY_MODULE_STREAM,
+
+  N_PROPS
+};
+
+static GParamSpec *properties[N_PROPS];
+
+ModulemdEol *
+modulemd_eol_new (guint64 mdversion,
+                  guint64 modified,
+                  const gchar *module_name,
+                  const gchar *module_stream,
+                  const gchar *message)
+{
+  // clang-format off
+  return g_object_new (MODULEMD_TYPE_EOL,
+                       "mdversion", mdversion,
+                       "modified", modified,
+                       "module-name", module_name,
+                       "module-stream", module_stream,
+                       "message", message,
+                       NULL);
+  // clang-format on
+}
+
+
+ModulemdEol *
+modulemd_eol_copy (ModulemdEol *self)
+{
+  g_autoptr (ModulemdEol) o = NULL;
+
+  g_return_val_if_fail (MODULEMD_IS_EOL (self), NULL);
+
+
+  o = modulemd_eol_new (modulemd_eol_get_mdversion (self),
+                        modulemd_eol_get_modified (self),
+                        modulemd_eol_get_module_name (self),
+                        modulemd_eol_get_module_stream (self),
+                        modulemd_eol_get_message (self));
+  modulemd_eol_set_module_context (o, modulemd_eol_get_module_context (self));
+  modulemd_eol_set_reset (o, modulemd_eol_get_reset (self));
+  modulemd_eol_set_eol_date (o, modulemd_eol_get_eol_date (self));
+  modulemd_eol_set_obsoleted_by_module_name (
+    o, modulemd_eol_get_obsoleted_by_module_name (self));
+  modulemd_eol_set_obsoleted_by_module_stream (
+    o, modulemd_eol_get_obsoleted_by_module_stream (self));
+
+  return g_steal_pointer (&o);
+}
+
+gboolean
+modulemd_eol_validate (ModulemdEol *self, GError **error)
```

Comment:
Just for the record, @kontura found the issues in my test data and knows about them:

You should validate the module (==name),  stream, and context fields.
Also obsoleted_by / {module, stream} should be validated.
I managed to use ':' character in the stream and it's definitely a forbidden character (NSVCA separator).

DNF also did not report a missing mandatory field (modified), but that might be due to that part of code being in progress.